File: StudyGamePrograming06/RacingCar.cpp
#include "RacingCar.h"
#include "MeshComponent.h"
#include "Game.h"
#include "Renderer.h"
#include "MoveComponent.h"

RacingCar::RacingCar(Game* game) :Actor(game)
{
	SetPosition(Vector3(500.0f, 700.0f, -100.0f));
	SetScale(1.0f);
	MeshComponent* mc = new MeshComponent(this);
	mc->SetMesh(game->GetRenderer()->GetMesh("Assets/RacingCar.gpmesh"));
	mMoveComp = new MoveComponent(this);
}

void RacingCar::UpdateActor(float deltaTime)
{
	float w = Math::Pi * 0.25f;
	/*
	float r = Math::Sqrt(GetPosition().x * GetPosition().x + GetPosition().y * GetPosition().y);
	float x = GetPosition().x * Math::Cos(w * deltaTime) - GetPosition().y * Math::Sin(w * deltaTime);
	float y = GetPosition().y * Math::Cos(w * deltaTime) + GetPosition().x * Math::Sin(w * deltaTime);
	SetPosition(Vector3(x, y, -100.0f));
	*/
	Quaternion q = GetRotation();
	Quaternion inc = Quaternion(Vector3::UnitZ, w * deltaTime);
	q = Quaternion::Concatenate(inc, q);
	SetRotation(q);
}

RacingCar2::RacingCar2(Game* game) :Actor(game)
{
	SetPosition(Vector3(0.0f, 700.0f, -100.0f));
	SetScale(1.0f);
	MeshComponent* mc = new MeshComponent(this);
	mc->SetMesh(game->GetRenderer()->GetMesh("Assets/RacingCar2.gpmesh"));
	mMoveComp = new MoveComponent(this);
}

void RacingCar2::UpdateActor(float deltaTime)
{
	float w = Math::Pi * 0.25f;
	/*
	float r = Math::Sqrt(GetPosition().x * GetPosition().x + GetPosition().y * GetPosition().y);
	float x = GetPosition().x * Math::Cos(w * deltaTime) - GetPosition().y * Math::Sin(w * deltaTime);
	float y = GetPosition().y * Math::Cos(w * deltaTime) + GetPosition().x * Math::Sin(w * deltaTime);
	SetPosition(Vector3(x, y, -100.0f));
	*/
	Quaternion q = GetRotation();
	Quaternion inc = Quaternion(Vector3::UnitZ, w * deltaTime);
	q = Quaternion::Concatenate(inc, q);
	SetRotation(q);
}

RacingCar3::RacingCar3(Game* game) :Actor(game)
{
	SetPosition(Vector3(-500.0f, 700.0f, -100.0f));
	SetScale(1.0f);
	MeshComponent* mc = new MeshComponent(this);
	mc->SetMesh(game->GetRenderer()->GetMesh("Assets/RacingCar3.gpmesh"));
	mMoveComp = new MoveComponent(this);
}

void RacingCar3::UpdateActor(float deltaTime)
{
	float w = Math::Pi * 0.25f;
	/*
	float r = Math::Sqrt(GetPosition().x * GetPosition().x + GetPosition().y * GetPosition().y);
	float x = GetPosition().x * Math::Cos(w * deltaTime) - GetPosition().y * Math::Sin(w * deltaTime);
	float y = GetPosition().y * Math::Cos(w * deltaTime) + GetPosition().x * Math::Sin(w * deltaTime);
	SetPosition(Vector3(x, y, -100.0f));
	*/
	Quaternion q = GetRotation();
	Quaternion inc = Quaternion(Vector3::UnitZ, w * deltaTime);
	q = Quaternion::Concatenate(inc, q);
	SetRotation(q);
}

File: StudyGamePrograming06_VB/RacingCar.vb
Imports OpenTK.Mathematics
Imports OpenTK.Windowing.GraphicsLibraryFramework
Public Class RacingCar
    Inherits Actor
    'public
    Sub New(ByRef game As Game)
        MyBase.New(game)
        SetPosition(New Vector3(0.0, 500.0, -100.0))
        'Dim q = Quaternion.FromAxisAngle(Vector3.UnitZ, Math.PI)
        'SetRotation(q)
        SetScale(1.0)
        Dim mc = New MeshComponent(Me)
        mc.SetMesh(game.GetRenderer.GetMesh("Assets/RacingCar.gpmesh"))
        mMoveComp = New MoveComponent(Me, 20)
    End Sub
    Public Overrides Sub UpdateActor(deltaTime As Single)
        MyBase.UpdateActor(deltaTime)
        Dim w As Single = Math.PI / 2
        Dim r As Single = Math.Sqrt(GetPosition.X * GetPosition.X + GetPosition.Y * GetPosition.Y)
        Dim x As Single = GetPosition.X * Math.Cos(w * deltaTime) + GetPosition.Y * Math.Sin(w * deltaTime)
        Dim y As Single = GetPosition.Y * Math.Cos(w * deltaTime) - GetPosition.X * Math.Sin(w * deltaTime)
        SetPosition(New Vector3(x, y, -100.0))
        Dim q As Quaternion = GetRotation()
        Dim inc As Quaternion = Quaternion.FromAxisAngle(Vector3.UnitZ, -w * deltaTime)
        q = Quaternion.Multiply(inc, q)
        SetRotation(q)
    End Sub
    'private
    Dim mMoveComp As MoveComponent
End Class
